### PR TITLE
Add codec integration test

### PR DESCRIPTION
**Pull request by making**
and fix codecs so that codecs can encode and decode data against real PostgreSQL.

### CODE REVIEW

**Comment by nebhale** [comment]
@davecramer I would have guessed that I built these codecs based on the JDBC driver, but I've just gone through and that certainly doesn't seem to be the case, so I have no idea where they came from.  If you've got the time, can you please give this a review and ensure that it matches up with your understanding of how encoding and decoding should happen for these types?

I also have a specific question about UUID changing from binary to text.  Could that be related to our discussion about connection configuration preferring binary or text?

**Comment by davecramer** [comment]
I haven't had time to go through all of it but as I am reviewing something on JDBC this jumps out as something that has to be dealt with.

https://github.com/pgjdbc/pgjdbc/blob/08631ccdabdb8ba6d52f398e2b0b46a9cf0cafbf/pgjdbc/src/main/java/org/postgresql/PGStatement.java#L21-L24

these are MIN and MAX values for PostgreSQL times and dates..

FYI, times and dates are the cause of many issues, mostly due to assumptions made about time zones

**Comment by nebhale** [comment]
Those problems with dates and times are one of the reasons I want to convert straight to the rich Java 8 types which take these things into account, where we can.  That special case seems to be centered around `java.sql.Date` and its contract.  With what you know there, do you think it pertains to `java.time` types as well?  Is the problem actually a mismatch between _ANSI/ISO SQL_ date/time and Java time?

**Comment by nebhale** [comment]
@davecramer Thanks for the clarification.

File: src/main/java/io/r2dbc/postgresql/codec/ZonedDateTimeCodec.java [hunk 1]
```diff
@@ -56,14 +56,14 @@ boolean doCanDecode(Format format, PostgresqlObjectId type) {
...
-        return ZonedDateTime.parse(ByteBufUtils.decode(byteBuf));
+        return PostgresqlDateTimeFormatter.INSTANCE.parse(ByteBufUtils.decode(byteBuf), ZonedDateTime::from);
     }
 
     @Override
     Parameter doEncode(ZonedDateTime value) {
         Objects.requireNonNull(value, "value must not be null");
 
-        ByteBuf encoded = ByteBufUtils.encode(this.byteBufAllocator, value.toString());
+        ByteBuf encoded = ByteBufUtils.encode(this.byteBufAllocator, value.toOffsetDateTime().toString());
```
Comment (nebhale) [inline src/main/java/io/r2dbc/postgresql/codec/ZonedDateTimeCodec.java]: Trying to understand this change.  Would this be considered a loss in fidelity?

Comment (davecramer) [reply src/main/java/io/r2dbc/postgresql/codec/ZonedDateTimeCodec.java]: Ya, I think so. Take the case where you read it in UTC and then write it back in PST or whatever different timezone. 

IIRC the JDBC driver sets the timezone for the session so we know what it is, and we don't allow the user to change it.

**Comment by making** [comment]
> I also have a specific question about UUID changing from binary to text. Could that be related to our discussion about connection configuration preferring binary or text?

When I tried the integration test with `BINARY` format, I got the following error

```
java.lang.AssertionError: expectation "expectNext(a3b245fe-ac1b-4e57-bef0-be257952147c)" failed (expected: onNext(a3b245fe-ac1b-4e57-bef0-be257952147c); actual: onError(java.lang.IllegalArgumentException: Cannot decode value of type java.util.UUID))

	at reactor.test.ErrorFormatter.assertionError(ErrorFormatter.java:105)
	at reactor.test.ErrorFormatter.failPrefix(ErrorFormatter.java:94)
	at reactor.test.ErrorFormatter.fail(ErrorFormatter.java:64)
	at reactor.test.ErrorFormatter.failOptional(ErrorFormatter.java:79)
	at reactor.test.DefaultStepVerifierBuilder.lambda$addExpectedValue$10(DefaultStepVerifierBuilder.java:487)
	at reactor.test.DefaultStepVerifierBuilder$SignalEvent.test(DefaultStepVerifierBuilder.java:2105)
	at reactor.test.DefaultStepVerifierBuilder$DefaultVerifySubscriber.onSignal(DefaultStepVerifierBuilder.java:1401)
	at reactor.test.DefaultStepVerifierBuilder$DefaultVerifySubscriber.onExpectation(DefaultStepVerifierBuilder.java:1349)
	at reactor.test.DefaultStepVerifierBuilder$DefaultVerifySubscriber.onError(DefaultStepVerifierBuilder.java:1023)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:185)
	at reactor.core.publisher.MonoFlatMap$FlatMapInner.onError(MonoFlatMap.java:251)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:185)
	at reactor.core.publisher.MonoFlatMap$FlatMapInner.onError(MonoFlatMap.java:251)
	at reactor.core.publisher.MonoSingle$SingleSubscriber.onError(MonoSingle.java:141)
	at reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:126)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:106)
	at reactor.core.publisher.FluxZip$ZipCoordinator.drain(FluxZip.java:735)
	at reactor.core.publisher.FluxZip$ZipInner.onNext(FluxZip.java:894)
	at reactor.core.publisher.FluxRepeatPredicate$RepeatPredicateSubscriber.onNext(FluxRepeatPredicate.java:78)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:114)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1476)
	at reactor.core.publisher.MonoProcessor.subscribe(MonoProcessor.java:457)
	at reactor.core.publisher.MonoMap.subscribe(MonoMap.java:55)
	at reactor.core.publisher.Mono.subscribe(Mono.java:3590)
	at reactor.core.publisher.FluxRepeatPredicate$RepeatPredicateSubscriber.resubscribe(FluxRepeatPredicate.java:112)
	at reactor.core.publisher.MonoRepeatPredicate.subscribe(MonoRepeatPredicate.java:47)
	at reactor.core.publisher.Flux.subscribe(Flux.java:7734)
	at reactor.core.publisher.FluxZip$ZipCoordinator.subscribe(FluxZip.java:579)
	at reactor.core.publisher.FluxZip.handleBoth(FluxZip.java:326)
	at reactor.core.publisher.FluxZip.handleArrayMode(FluxZip.java:267)
	at reactor.core.publisher.FluxZip.subscribe(FluxZip.java:136)
	at reactor.core.publisher.FluxMap.subscribe(FluxMap.java:62)
	at reactor.core.publisher.MonoSingle.subscribe(MonoSingle.java:58)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:150)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1476)
	at reactor.core.publisher.MonoSingle$SingleSubscriber.onComplete(MonoSingle.java:171)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onComplete(FluxMapFuseable.java:144)
	at reactor.core.publisher.FluxWindowPredicate$WindowPredicateMain.checkTerminated(FluxWindowPredicate.java:487)
	at reactor.core.publisher.FluxWindowPredicate$WindowPredicateMain.drainLoop(FluxWindowPredicate.java:435)
	at reactor.core.publisher.FluxWindowPredicate$WindowPredicateMain.drain(FluxWindowPredicate.java:379)
	at reactor.core.publisher.FluxWindowPredicate$WindowPredicateMain.onComplete(FluxWindowPredicate.java:290)
	at reactor.core.publisher.MonoFlatMapMany$FlatMapManyInner.onComplete(MonoFlatMapMany.java:248)
	at reactor.core.publisher.FluxWindowPredicate$WindowFlux.checkTerminated(FluxWindowPredicate.java:715)
	at reactor.core.publisher.FluxWindowPredicate$WindowFlux.drainRegular(FluxWindowPredicate.java:609)
	at reactor.core.publisher.FluxWindowPredicate$WindowFlux.drain(FluxWindowPredicate.java:695)
	at reactor.core.publisher.FluxWindowPredicate$WindowFlux.onComplete(FluxWindowPredicate.java:761)
	at reactor.core.publisher.FluxWindowPredicate$WindowPredicateMain.onNext(FluxWindowPredicate.java:241)
	at reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:113)
	at reactor.core.publisher.FluxHandle$HandleConditionalSubscriber.onNext(FluxHandle.java:320)
	at reactor.core.publisher.FluxHandle$HandleConditionalSubscriber.onNext(FluxHandle.java:320)
	at reactor.core.publisher.FluxHandle$HandleConditionalSubscriber.onNext(FluxHandle.java:320)
	at reactor.core.publisher.FluxHandle$HandleConditionalSubscriber.onNext(FluxHandle.java:320)
	at reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onNext(FluxPeekFuseable.java:826)
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.innerNext(FluxConcatMap.java:275)
	at reactor.core.publisher.FluxConcatMap$ConcatMapInner.onNext(FluxConcatMap.java:849)
	at reactor.core.publisher.FluxGenerate$GenerateSubscription.next(FluxGenerate.java:169)
	at io.r2dbc.postgresql.message.backend.BackendMessageDecoder.lambda$decode$1(BackendMessageDecoder.java:133)
	at reactor.core.publisher.FluxGenerate$GenerateSubscription.slowPath(FluxGenerate.java:262)
	at reactor.core.publisher.FluxGenerate$GenerateSubscription.request(FluxGenerate.java:204)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.set(Operators.java:1849)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onSubscribe(Operators.java:1723)
	at reactor.core.publisher.FluxGenerate.subscribe(FluxGenerate.java:83)
	at reactor.core.publisher.Flux.subscribe(Flux.java:7734)
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.drain(FluxConcatMap.java:442)
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.onNext(FluxConcatMap.java:244)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onNext(FluxPeek.java:192)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:114)
	at reactor.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:211)
	at reactor.netty.channel.FluxReceive.onInboundNext(FluxReceive.java:327)
	at reactor.netty.channel.ChannelOperations.onInboundNext(ChannelOperations.java:310)
	at reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:141)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:628)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:528)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:482)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:442)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884)
	at java.base/java.lang.Thread.run(Thread.java:834)
	Suppressed: java.lang.IllegalArgumentException: Cannot decode value of type java.util.UUID
		at io.r2dbc.postgresql.codec.DefaultCodecs.decode(DefaultCodecs.java:91)
		at io.r2dbc.postgresql.PostgresqlRow.get(PostgresqlRow.java:82)
		at io.r2dbc.postgresql.codec.CodecIntegrationTest.lambda$testUuid$82(CodecIntegrationTest.java:557)
		at io.r2dbc.postgresql.PostgresqlResult.lambda$map$0(PostgresqlResult.java:69)
		at reactor.function.TupleUtils.lambda$function$7(TupleUtils.java:143)
		at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:100)
		... 73 more
```

And confirmed the uuid came as a `TEXT`.



I may miss something. Is there any way to take an uuid as `BINARY` ?

**Comment by making** [comment]
rebased

**Comment by nebhale** [comment]
@making Merging in now as-is.  I think we'll need a couple more passes over the Codecs as time goes on, but this is a strong improvement as it exists today, so worth it to go now.  Please take a quick look at the polishing commit, specifically `CodecIntegrationTest` which is 1/3 as long now.

**Comment by making** [comment]
Thanks for polishing!

**Pull request closed without merging**